Move "Make Primary" to domains list

umurkontaci opened the pull request:
Fixes #2531.

This add the Make Primary button to Domains -> List. (It does not remove the old way – let's watch adoption first).

### Screenshots

#### Primary Domain Mode Disabled (default)
![image](https://cloud.githubusercontent.com/assets/991022/13654533/cc627f80-e60d-11e5-9989-5816f143bedb.png)

#### Primary Domain Mode Enabled
![image](https://cloud.githubusercontent.com/assets/991022/13654550/e59d16a4-e60d-11e5-8338-8674b35ad86e.png)

#### Click on domain
![image](https://cloud.githubusercontent.com/assets/991022/13654562/f92de810-e60d-11e5-84d9-510500a71095.png)

#### Successful change
![image](https://cloud.githubusercontent.com/assets/991022/13654579/087f49b2-e60e-11e5-9669-70e356cc30ec.png)

#### Successful undo

![image](https://cloud.githubusercontent.com/assets/991022/13676347/531fd220-e69a-11e5-9540-970064aba944.png)


#### Fail change
![image](https://cloud.githubusercontent.com/assets/991022/13654737/d6a0a1ba-e60e-11e5-91f1-05777ad4034a.png)


### Testing
- Go to Domains -> List
- Click "Change Primary"
- Click any of the domains
- Wait
- Confirm you've changed the primary domain
- Confirm you the URL changes at the same time with the success notice
- Confirm you can navigate the site without problems (back button will break for old URLs – it's a known issue without a feasible fix)
- Confirm undo works.
- Changing primary domain name is a fragile thing to do since we rely on URLs to determine the active site and we are actively changing the canonical URL. So look out for stuff.

### Things to look for
- Event names for tracking

/cc: @klimeryk @aidvu 
/cc: @mikeshelton1503 @breezyskies

klimeryk commented on client/lib/domains/reducer.js:
```diff
@@ -82,6 +83,16 @@ function reducer( state, payload ) {
 			} );
 
 		case UpgradesActionTypes.PRIMARY_DOMAIN_SET_COMPLETED:
+			return updateSiteState( state, siteId, {
```
This is a lot of logic - while it could be a bit simplified, how about a different approach? We can pass any parameters we want along with the action. I think it should be possible to pass the old and new `domainName`s - that way, we wouldn't have to iterate over the whole array, just update those two entries. While this will add a bit of logic on the view, I feel it's a cleaner overall approach.

umurkontaci replied:
It looks a bit complicated but it's just reducer's job. I don't think storing the previous name is a good approach. If we can deduce a data from the current values without adding any extra data, then we should do that. Passing in the previous domain name while you know the previous domain name feels a bit wrong to me.

Besides, even we have the previous domain name, we should see iterate the array to find the match for the domain in the array, and we still have to sort the array so the primary domain is on top.

klimeryk replied:
I guess you're right - kinda like we shouldn't add unnecessary things to a component's state.
OK, so, instead, what do you think about changing:
```
if ( action.domainName === domain.name ) {
	return Object.assign( {}, domain, { isPrimary: true } );
}
return Object.assign( {}, domain, { isPrimary: false } );
```
to just:
```
return Object.assign( {}, domain, { isPrimary: action.domainName === domain.name } );
```

umurkontaci replied:
That works!

umurkontaci replied:
:+1:

klimeryk commented on client/lib/upgrades/actions/domain-management.js:
```diff
@@ -30,12 +39,13 @@ function setPrimaryDomain( siteId, domainName, onComplete ) {
 			} );
 
-			return onComplete( error, data );
+			onComplete( error, data );
```
Are you sure we don't want `return` here?

umurkontaci replied:
We do. Fixed

drw158 commented:
The Successful Undo is confusing. The two screenshots look the same. Intentional?

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -46,26 +65,183 @@ const List = React.createClass( {
+				Object.assign( props, {
+					text: <span key="text">{ this.translate( 'Primary Domain changed: All domains will redirect to' +
+						' {{em}}%(domainName)s' +
+						'{{/em}}.', { args: { domainName }, components: { em: <em/> } } ) }</span>,
```
This seems like a weird place to break the string? `{{em}}%(domainName)s{{/em}}.` can definitely be on one line.

umurkontaci replied:
I blame Webstorm.

umurkontaci replied:
:+1:

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -46,26 +65,183 @@ const List = React.createClass( {
+			case PRIMARY_DOMAIN_REVERT_SUCCESS:
+				Object.assign( props, {
+					text: <span key="text">{ this.translate( 'Primary Domain change reverted: All domains will now redirect to' +
+						' {{em}}' +
```
Same issue with breaking the string.

umurkontaci replied:
:+1:

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -46,26 +65,183 @@ const List = React.createClass( {
+				className="domain-management-list__change-primary-button"
+				onClick={ this.enableChangePrimaryDomainMode }>
+				{ this.translate( 'Change Primary', { context: 'Button label for changing primary domain' } ) }
+				</Button>
```
The indentation of this tag is, like, _completely_ wrong, it throws off the Feng shui of the whole function!
(I hope you realize I'm 99% joking).

umurkontaci replied:
Meh, this is what happens when I use 2 space tabs instead of four.

umurkontaci replied:
:+1:

klimeryk replied:
Just noting that this is still mis-indented - I don't really care about this that much, but I'm bringing it up because maybe you forgot to push a change with this. And/or, more importantly, possibly something else.

umurkontaci commented:
> The Successful Undo is confusing. The two screenshots look the same. Intentional?

Ah no, they have a different messaging actually, the screenshot is wrong. Updated the post

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -46,26 +65,183 @@ const List = React.createClass( {
+		this.setState( { notice: null } );
+	},
+
+	notice() {
```
This file seems long<i>-ish</i> to me - maybe we can extract some stuff to a separate components. For example, this notice function could become a nice, stateless React component - we love those! (no sarcasm intended)

umurkontaci replied:
I was thinking about doing that but couldn't think of an quick and easy way. Functional components might be the answer!

klimeryk commented on client/my-sites/upgrades/domain-management/list/item.jsx:
```diff
@@ -9,29 +9,95 @@ import React from 'react';
 import DomainPrimaryFlag from 'my-sites/upgrades/domain-management/components/domain/primary-flag';
 import { type as domainTypes } from 'lib/domains/constants';
+import Spinner from 'components/spinner';
+import classNames from 'classnames';
```
I think this is an external dependency :)

umurkontaci replied:
:+1:

klimeryk commented on client/my-sites/upgrades/domain-management/list/item.jsx:
```diff
@@ -9,29 +9,95 @@ import React from 'react';
+import Spinner from 'components/spinner';
+import classNames from 'classnames';
 
 var Gridicon = require( 'components/gridicon' );
```
This could be moved to the `import` section above (not introduced by this PR, but it stands out).

umurkontaci replied:
![image](https://cloud.githubusercontent.com/assets/991022/13676928/a4794226-e69c-11e5-96a2-1f4e975948e9.png)

umurkontaci replied:
:+1:

klimeryk commented:
For some reason, when I click "Change Primary" and then a domain I want to change to, I get this error in console:
```
warning.js:45 Warning: Failed propType: Invalid prop `isSingle` supplied to `PublishMenu`. Check the render method of `MySitesSidebar`.
```

klimeryk commented:
I think we should hide the "right pointing arrows" (aka `chevron-right`) when we're in the "setting primary domain" mode. Otherwise, it looks like you can navigate somewhere by clicking on it. @mikeshelton1503 @breezyskies: thoughts? :)

klimeryk commented:
Just wanted to add that I really enjoy reading your code, @umurkontaci (despite my nitpicking). It's always full of cool and useful patterns and I always learn something new by doing code reviews of your PRs. So thank you and keep them coming :bow:

umurkontaci commented:
> For some reason, when I click "Change Primary" and then a domain I want to change to, I get this error in console:

I think it's a bug in `PublishMenu` relying on `sites.selected` instead of `sites.getSelectedSite().slug`. I can fix that, but it seems out of scope of this PR. I'm not sure if it's actually broken though, but if that's the case I'm happy with holding this PR and fixing the bugs before merging this.

> I think we should hide the "right pointing arrows"

Agreed

> Failed propType: Required prop `settingPrimaryDomain` was not specified in `Header`. Check the render method of `MappedDomain`.

I guess these are new errors where you pass an extra prop that's not defined? I'll double check, in the meantime, could you also check the components behave normally after the change?



> Just wanted to add that I really enjoy reading your code, @umurkontaci (despite my nitpicking). It's always full of cool and useful patterns and I always learn something new by doing code reviews of your PRs. So thank you and keep them coming

Oh thanks! That is great to hear!

klimeryk commented:
> I think it's a bug in PublishMenu relying on sites.selected instead of sites.getSelectedSite().slug. I can fix that, but it seems out of scope of this PR. I'm not sure if it's actually broken though, but if that's the case I'm happy with holding this PR and fixing the bugs before merging this.

If it's a simple fix like that then either create a quick PR for it or incorporate it into this PR (as a separate commit, of course, to add a meaningful comment for it).

> > Failed propType: Required prop settingPrimaryDomain was not specified in Header. Check the render method of MappedDomain.

> I guess these are new errors where you pass an extra prop that's not defined? I'll double check, in the meantime, could you also check the components behave normally after the change?

Just checked again and these show even if you don't change the primary domain first. Just navigate to the domains list and click on any of them (doesn't look like anything actually breaks, though).

umurkontaci commented:
> Just checked again and these show even if you don't change the primary domain first. Just navigate to the domains list and click on any of them (doesn't look like anything actually breaks, though).

I think it's just that we are passing extra props to those components, and if they have `propTypes` defined, they complain about it. I'll see to remove the messages.

umurkontaci commented on client/my-sites/upgrades/domain-management/edit/card/header/index.jsx:
```diff
@@ -35,7 +35,7 @@ const Header = React.createClass( {
 					domain={ domain }
 					selectedSite={ this.props.selectedSite }
-					settingPrimaryDomain={ this.props.settingPrimaryDomain } />
+					settingPrimaryDomain={ this.props.primaryDomainModeEnabled } />
```
This shouldn't have changed.

umurkontaci replied:
:+1:

umurkontaci commented on client/my-sites/upgrades/domain-management/edit/card/header/primary-domain-button.jsx:
```diff
@@ -34,7 +34,7 @@ const PrimaryDomainButton = React.createClass( {
 
 		if ( domain && ! domain.isPrimary ) {
-			if ( this.props.settingPrimaryDomain ) {
+			if ( this.props.primaryDomainModeEnabled ) {
```
This shouldn't have changed.

umurkontaci replied:
:+1:

breezyskies commented:
> I think we should hide the "right pointing arrows" (aka chevron-right) when we're in the "setting primary domain" mode. Otherwise, it looks like you can navigate somewhere by clicking on it. 

Agreed, good suggestion.

umurkontaci commented:
Thanks @mikeshelton1503 !

I always get confused when implementing these type of changes to figure out the color and the size since they seem to vary a lot. It'd be great to have those written somewhere in the issue or a general guideline (could be they are already and I just don't know it). Otherwise it's trial and error, which is not really fun. I'll update the CSS for those.


> In what cases will this message display

The generic message is "Something went wrong and we couldn't change your Primary Domain."
This message showed up because I tried to set my primary domain to a domain I don't own. So, even though the message could be improved, it's technically correct :)




> Let's try something a little warmer like: "No worries, your primary domain has been set back to testestest.me."

Will do!

umurkontaci commented:
> I know, I'm sorry I should have provided that for you. It's something we plan to start doing for projects like this.

No worries! I was just for future reference and glad you're on top of it.

umurkontaci commented:
> I noticed that sometimes it seemed to take a really long time to process the change (the loading spinner kept on spinning). I then discovered that if I switched tabs (clicked on another browser tab briefly then went back) it would trigger the change to happen right after I switched back. Any ideas? Does that make any sense?

I don't think there's any reason for that to happening. It does take long, because after setting the primary domain, we need fetch sites and the domains for the site again. Currently there's no way around it since we're changing the canonical URL and things are fragile.

umurkontaci commented on client/my-sites/upgrades/domain-management/edit/index.jsx:
```diff
@@ -36,7 +36,7 @@ const Edit = React.createClass( {
 					domain={ domain }
 					selectedSite={ this.props.selectedSite }
-					settingPrimaryDomain={ this.props.domains.settingPrimaryDomain } />
+					settingPrimaryDomain={ this.props.domains.primaryDomainModeEnabled } />
```
:+1:

umurkontaci commented:
> Next up, if I click on my newly changed primary domain, I get these:

:+1:

umurkontaci commented:
> I think we should hide the "right pointing arrows"

:+1:

umurkontaci commented:
> below 460px the typeface used for the domain name changes to sans-serif.
> I mistakenly set the font-size of domain description text to 11px, however we are trying to standardize font-sizes used, so let's make it 12px

Both wasn't introduced by this PR, but was a quick win so I fixed it. :+1: 


> below 460px the font-size for "setting primary domain" changes to 12px which is bigger than the 11px on larger screens. Let's make it all 12px to keep inline with the above change.
:+1:

umurkontaci commented:
> If it's a simple fix like that then maybe you can either create a quick PR for it or just incorporate it into this PR (as a separate commit, of course, to add a meaningful comment for it).

:+1: Fixed.

I also moved the notice logic to a different component.

Sorry for all the email spam you got, but everything should've been finished by now. @klimeryk care for another test?

klimeryk commented on client/my-sites/sidebar/sidebar.jsx:
```diff
@@ -66,12 +66,12 @@ module.exports = React.createClass( {
 
 	isSingle: function() {
-		return this.props.sites.selected || this.props.sites.get().length === 1;
+		return !! ( this.props.sites.getSelectedSite() || this.props.sites.get().length === 1 );
```
Nice! :+1:

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -82,16 +209,78 @@ const List = React.createClass( {
+			// user clicked the current primary domain
+			this.setState( {
+				primaryDomainModeEnabled: false
+			} );
```
I think you meant to `return;` here. Otherwise this still tries to re-set the current domain as the primary one. Nothing breaks or anything, but obviously if we can skip the request to the server all the better :) (which was probably your intention here)

umurkontaci replied:
:+1:

klimeryk commented on client/my-sites/upgrades/domain-management/list/item.jsx:
```diff
@@ -9,29 +10,97 @@ import React from 'react';
+		enableSelection: React.PropTypes.bool,
+		onClick: React.PropTypes.func.isRequired,
+		onSelect: React.PropTypes.func.isRequired,
+		isSelected: React.PropTypes.bool
```
You're missing `selectionIndex` in `propTypes` :)

umurkontaci replied:
:+1:

klimeryk commented on client/my-sites/upgrades/domain-management/style.scss:
```diff
@@ -65,6 +65,24 @@
+
+.domain-management-list-item {
+	&.busy {
+		background-color: #f3f6f8;
```
This is `$gray-light` :)

umurkontaci replied:
Did you know that off the top of your head?

umurkontaci replied:
:+1:

klimeryk replied:
I wish... :disappointed: Nah, I just `grep`ed (or actually `ag`ed) for this value in our `assets` knowing that Mike probably used a color already defined in our palette somewhere.

klimeryk commented on client/my-sites/upgrades/domain-management/style.scss:
```diff
@@ -65,6 +65,24 @@
+		background-color: #f3f6f8;
+
+		.domain-management-list-item__title {
+			color: #bcc4ca;
```
Maybe just use the same color (according to my eyes, they _should_ be the same color) as `domain-management-list-item__meta` - that way you won't use a "raw" color value and you can merge it into one rule:
```
.domain-management-list-item__title,
.domain-management-list-item__meta {
	color: lighten( $gray, 20% );
}
```

umurkontaci replied:
How do you know those colors? :+1:

klimeryk replied:
Here I just noticed that you were using `lighten( $gray, 20% )` for `.domain-management-list-item__meta`, but on the view it looked exactly like `#bcc4ca` in `.domain-management-list-item__title` above it (in the view). I didn't bother to check it with a color dropper or anything, just hoped it's actually the same :laughing:

klimeryk commented:
Thanks for the fixes! Basically all of the old issues were fixed (apart from that misindented `Button`), but I've noticed some new issue, especially the one with "allowing" setting the current primary domain as the primary domain again via an unnecessary server request.

umurkontaci commented:
> I've noticed some new issue, especially the one with "allowing" setting the current primary domain as the primary domain again via an unnecessary server request.

That's just fixed :)

klimeryk commented on client/my-sites/upgrades/domain-management/list/index.jsx:
```diff
@@ -82,16 +209,79 @@ const List = React.createClass( {
+		if ( domain.name === currentPrimary ) {
+			// user clicked the current primary domain
+			this.setState( {
+				primaryDomainModeEnabled: false
```
Aaargh, missed this one - this should be `changePrimaryDomainModeEnabled` I think. Especially since `primaryDomainModeEnabled` is never referenced again ;)

umurkontaci commented:
@klimeryk I've fixed everything, added some unit tests to the component. 

@blowery I've added some mocks and added a second way to use `useSinon` without depending on mocha's `this` context. Would welcome a review on those :)

blowery commented on test/test/helpers/use-sinon/index.js:
```diff
@@ -43,15 +43,17 @@ export function useFakeTimers( now = 0, clockCallback = noop ) {
 
 	before( function() {
-		this.sandbox = sinon.sandbox.create( config );
+		wrapper.sandbox = this.sandbox = sinon.sandbox.create( config );
```
why a prop on wrapper rather than just the sandbox itself?

umurkontaci replied:
copy by value? 
`before` and `after` might be called after the function returns in which case we'll return `undefined` or something out of sync with the actual sandbox.

We can return a getter function, but I thought it might be too confusing.

blowery replied:
oh, right, duh. 

what's the advantage over just using `sinon.test` or `sinon.testCase` instead? See http://sinonjs.org/docs/#sinon-test

umurkontaci replied:
It also binds its props to `this`? I think in order to able to use `test`, we need to use one ES5 function to pass to `sinon.test` and the rest of the functions must be arrow functions to retain the value of `this`. Seems more fragile?

blowery commented on test/test/helpers/mocks/README.md:
```diff
@@ -0,0 +1,22 @@
+const mockComponentClasses = require( 'test/helpers/mocks/component-classes' );
+const mockery = require( 'mockery' );
+describe( 'my-module', function() {
+  before( function() {
```
what about a usage more like this:
```
useMockery = ...;
mockComponentClasses = ...;

describe(..., function() {
  useMockery( mockery => {
    mockComponentClasses( mockery ) // mockery could even be optional here
    mockTip( mockery )
  } )
})
```
and then let useMockery take care of cleanup

umurkontaci replied:
I like it, but I'm kinda hesitant on implicit cleanups. Is there a case that we might want more granularity in terms of lifecycle? Can we ever want the component to have a different lifecycle than `useMockery`?

blowery replied:
> Can we ever want the component to have a different lifecycle than useMockery?

We can't have it extend past the use of mockery, which is why I like it nested. If someone needs to reset the mock between uses, they can require what's being mocked in a afterEach and do the work there.

umurkontaci replied:
Right. I'll switch to your suggestion :+1:

umurkontaci replied:
Done.

umurkontaci replied:
Done

umurkontaci commented:
Tests were never ending due to `data-poller` having active callbacks, I created a common mock for that as well.

klimeryk commented:
Everything works as expected and the code looks good to me :+1: 
Unless @blowery has any comments or issues with the new mocks and/or tests, this looks good to :ship:

blowery commented:
Looks good. Happy.
